walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,151 @@
+    List<TimeRange> occupiedTimeRanges = events
+      .stream()
+      .filter(event -> FindMeetingQuery.hasCommonAttendees(event, attendees))
+      .map(event -> event.getWhen())
+      .sorted(TimeRange.ORDER_BY_START)
+      .collect(Collectors.toList());
+    
+    List<TimeRange> result = getNewMeetingTimeRanges(occupiedTimeRanges, duration);
+
+    return result;
+  }
+
+  /**
+   * Returns whether the {@code event} has at least one common attendee as the {@code requestedAttendees}.
+   */
+  private static boolean hasCommonAttendees(Event event, Set<String> requestedAttendees) {
+    Set<String> eventAttendees = new HashSet<>(event.getAttendees());
+    for (String requestedAttendee : requestedAttendees) {
+      if (eventAttendees.contains(requestedAttendee)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+
+  /**
+   * Returns the times slots when a new meeting taking {@code newMeetingDuration} minutes could happen 
+   * in a day, considering the {@code occupiedTimeRanges}, which are sorted in ascending order by start.
+   * It uses a Greedy algorithm.
+   */
+  private static List<TimeRange> getNewMeetingTimeRanges(List<TimeRange> occupiedTimeRanges, 
+                                                         long newMeetingDuration) {
+    // Consider the cases when there are no constraints for the new meeting.
+    if (occupiedTimeRanges.isEmpty()) {
+      if (newMeetingDuration <= TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      } else {
+        return Collections.emptyList();
+      }
+    }
+
+    // Init the time slots list when the new meeting can take place. 
+    List<TimeRange> newMeetingTimeRanges = new ArrayList<>();
+
+    Iterator<TimeRange> occupiedTimeRangesIterator = occupiedTimeRanges.iterator();
+    TimeRange currentOccupiedTimeRange = occupiedTimeRangesIterator.next();
+
+    // Check for empty time slot before the first occupied time range
+    if (currentOccupiedTimeRange.start() > TimeRange.START_OF_DAY) {
+      TimeRange newTimeRange = 
+          TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currentOccupiedTimeRange.start(), false);
+
+      if (newTimeRange.duration() >= newMeetingDuration) {
+        newMeetingTimeRanges.add(newTimeRange);
+      }
+    }
+
+    int previousOccupiedTimeRangeEnd = currentOccupiedTimeRange.end();
+
+    while(occupiedTimeRangesIterator.hasNext()) {
+      currentOccupiedTimeRange = occupiedTimeRangesIterator.next();
+
+      if (previousOccupiedTimeRangeEnd < currentOccupiedTimeRange.start()) {
+        // Found an empty time slot: [ previousOccupiedTimeRangeEnd, currentOccupiedTimeRange.start() )
+        TimeRange newTimeRange = 
+            TimeRange.fromStartEnd(previousOccupiedTimeRangeEnd, currentOccupiedTimeRange.start(), false);
+
+        if (newTimeRange.duration() >= newMeetingDuration) {
+          newMeetingTimeRanges.add(newTimeRange);
+        }
```

Comment:
This code is almost the same with the one above. Shall we move it to a shared static method?

Reply:
I added this method

```java 
 /**
  * Returns whether {@code duration} has at least {@code minimumDuration} minutes.
  */
  private static boolean checkTimeRangeDuration(long duration, long minimumDuration) {
    return (duration >= minimumDuration);
  }
```

The thing that bothers me is that I cannot call by reference in Java as in C++, so I cannot update my array inside the method 😢